src/lib/homestore/shard_manager.cpp

```diff
@@ -61,183 +62,119 @@ ShardManager::Result< ShardInfo > HSHomeObject::_create_shard(pg_id pg_owner, ui
         LOGWARN("failed to create shard with non-exist pg [{}]", pg_owner);
         return folly::makeUnexpected(ShardError::UNKNOWN_PG);
     }
-    // TODO: will update to ReplDev when ReplDev on HomeStore is ready;
-    auto replica_set_var = _repl_svc->get_replica_set(fmt::format("{}", pg_owner));
-    if (std::holds_alternative< home_replication::ReplServiceError >(replica_set_var)) {
+
+    homestore::ReplicationService* replication_service =
+        (homestore::ReplicationService*)(&homestore::HomeStore::instance()->repl_service());
+    auto repl_dev = replication_service->get_replica_dev(pg.value().repl_dev_uuid);
+    if (!bool(repl_dev)) {
         LOGWARN("failed to get replica set instance for pg [{}]", pg_owner);
         return folly::makeUnexpected(ShardError::UNKNOWN_PG);
     }
 
-    auto replica_set = std::get< home_replication::rs_ptr_t >(replica_set_var);
     auto new_shard_id = generate_new_shard_id(pg_owner);
     auto create_time = get_current_timestamp();
     auto shard_info =
         ShardInfo(new_shard_id, pg_owner, ShardInfo::State::OPEN, create_time, create_time, size_bytes, size_bytes, 0);
     std::string create_shard_message = serialize_shard(Shard(shard_info));
-    // preapre msg header;
-    const uint32_t needed_size = sizeof(ReplicationMessageHeader) + create_shard_message.size();
-    auto buf = nuraft::buffer::alloc(needed_size);
-    uint8_t* raw_ptr = buf->data_begin();
-    ReplicationMessageHeader* header = new (raw_ptr) ReplicationMessageHeader();
-    header->message_type = ReplicationMessageType::SHARD_MESSAGE;
-    header->payload_size = create_shard_message.size();
-    header->payload_crc =
-        crc32_ieee(init_crc32, r_cast< const uint8_t* >(create_shard_message.c_str()), create_shard_message.size());
-    header->header_crc = header->calculate_crc();
-    raw_ptr += sizeof(ReplicationMessageHeader);
-    std::memcpy(raw_ptr, create_shard_message.c_str(), create_shard_message.size());
 
+    // round up msg size to homestore datasvc blk size
+    const auto datasvc_blk_size = homestore::HomeStore::instance()->data_service().get_blk_size();
+    auto msg_size = create_shard_message.size();
+    if (msg_size % datasvc_blk_size != 0) { msg_size = (msg_size / datasvc_blk_size + 1) * datasvc_blk_size; }
+    auto msg_buf = iomanager.iobuf_alloc(512, msg_size);
+    std::memset(r_cast< uint8_t* >(msg_buf), 0, msg_size);
+    std::memcpy(r_cast< uint8_t* >(msg_buf), create_shard_message.c_str(), create_shard_message.size());
+
+    // preapre msg header;
+    ReplicationMessageHeader header;
+    header.repl_group_id = pg.value().pg_info.id;
+    header.msg_type = ReplicationMessageType::CREATE_SHARD_MSG;
+    header.payload_size = msg_size;
+    header.payload_crc = crc32_ieee(init_crc32, r_cast< const uint8_t* >(msg_buf), msg_size);
+    header.header_crc = header.calculate_crc();
```

Comment:
Should we make it an internal crc, so that crc itself is also protected?

Reply:
One argument I had made to myself before I had made a similar comment was that we are not exposing the `how` for this message to HomeObject...so technically knowing that it's not some magic replicated spinning disk that just throws off random bits is not an assumption it should make? /shrug

Reply:
I think in this case, it is more of protection from memory corruption which can not be from inside. Do you think it is an overkill from doing that?

Reply:
Hi @yamingk , please check if the latest change is what you wants

Reply:
it is external crc now, e.g. crc itself is excluded in header.claculate_crc(). 
I am fine with this approach, given the magic in header can also provide some protection against memory corruption. But ideally an internal crc is with protection to the crc field itself. I will leave it to you to decide.

Reply:
ok, I think I got your point here, yes, it will be more reasonable if header_crc is included too. but if I do that, there is an extra place needed to save the crc of the whole header, right?

Reply:
no, crc itself can still be inside this data structure. 
you can refer to this if you decide to do similar (not a must): https://github.com/eBay/HomeStore/blob/master/src/lib/device/device.h#L50

Reply:
sure, will take a look and apply.